### PR TITLE
[Multitenancy]: Add PNI CRD

### PR DESCRIPTION
**Pull request by aggarwal0009**
**Reason for Change**:

New CRD PodNetworkInstance added to enable VNET multitenancy 

PNIs represent optional requirements, or behavior configurations for how we setup the pod networking. They should map 1:1 and follow the lifetime of a customer workload.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [ ] adds unit tests
- [x] relevant PR labels added

**Notes**:

### CODE REVIEW

File: crd/external/podnetworkinstance/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Pod Network Instance (PNI)
+
+PNIs represent optional requirements, or behavior configurations for how we setup the pod networking. They should map 1:1 and follow the lifetime of a customer workload.
+
+The object points to the PN for the delegated subnet to use and defines allocation requirements (e.g.: for IPs to reserve for pod endpoints). Orchestrator RP will map the deployments with these requirements to the PNI object through labels on the pod spec pointing to this object identifier.
```
Comment (miguelgoms) [inline crd/external/podnetworkinstance/README.md:5]: nit: expand PN to PodNetwork, "orchestrator RP will map" with "orchestrators can map"

Comment (aggarwal0009) [reply crd/external/podnetworkinstance/README.md:5]: updated

**Review by miguelgoms — DISMISSED** [review]

File: crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+type PodNetworkInstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []PodNetworkInstance `json:"items"`
+}
+
+// PodNetworkInstanceSpec defines the desired state of PodNetworkInstance
+type PodNetworkInstanceSpec struct {
+	// +kubebuilder:default=0
+	// +kubebuilder:validation:Optional
```
Comment (thatmattlong) [inline crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go:43]: not exactly sure what this annotation means but a PodNetwork is not an optional field of a PodNetworkInstance

File: crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	// pod network resource object name
+	PodNetwork string `json:"podnetwork,omitempty"`
+	// number of backend IP address to reserve for running pods
+	// +kubebuilder:default=0
+	PodIPReservationSize int `json:"podIPReservationSize"`
+}
+
+// PodNetworkInstanceStatus defines the observed state of PodNetworkInstance
+type PodNetworkInstanceStatus struct {
+	// +kubebuilder:validation:Optional
```
Comment (thatmattlong) [inline crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go:53]: same here, I wouldn't mark this optional

File: crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go (hunk 2 above)
Comment (aggarwal0009) [reply crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go:43]: hmm, just removing this annotation is not enough to make this field required. I would also need to drop omit empty from the field for it to be marked required in CRD
With omitempty in place this annotation has no effect.

Comment (aggarwal0009) [reply crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go:43]: so I dropped omitempty from podnetwork and it is now marked req in generated CRD.

File: crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go (hunk 3 above)
Comment (aggarwal0009) [reply crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go:53]: removed but it has no effect. Do I need to drop omitempty from PodIPAddresses field here? They can be empty initially right?

File: .github/workflows/crdgen.yaml [hunk 4]
```diff
@@ -26,9 +26,13 @@ jobs:
...
     - name: Regenerate MultitenantNetworkContainer CRD
       run: make -C crd/multitenantnetworkcontainer
+    - name: Regenerate PodNetworkInstance CRD
+      run: make -C crd/external/podnetworkinstance
     - name: Regenerate NodeInfo CRD
       run: make -C crd/nodeinfo
     - name: Regenerate MultitenantPodNetworkConfig CRD
       run: make -C crd/multitenantpodnetworkconfig
+    - name: Regenerate PodNetworkInstance CRD
+      run: make -C crd/external/podnetworkinstance
```
Comment (rbtr) [inline .github/workflows/crdgen.yaml:36]: duplicate steps in this workflow again

**Review by rbtr — CHANGES_REQUESTED** [review]

File: .github/workflows/crdgen.yaml (hunk 4 above)
Comment (aggarwal0009) [reply .github/workflows/crdgen.yaml:36]: removed

**Review by miguelgoms — DISMISSED** [review]

File: crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go (hunk 3 above)
Comment (thatmattlong) [reply crd/external/podnetworkinstance/api/v1alpha1/podnetworkinstance.go]: hm yeah let's keep it like you have it for now

**Review by thatmattlong — DISMISSED** [review]

**Comment by neaggarwMS** [comment]
Merging as Dual stack E2E is flaky. Pinging Ramiro to get that fixed.

**Pull request merged**